Implement YAML serialization (#4)

maxfadson opened the pull request:
I added an implementation of YAML serialization, wrote test cases and documentation. I used Jason's already implemented serializer as a basis, since YAML is very similar to it and is a more advanced format.

I suspect that not all cases were covered by tests, so I will add to the pull request as bugs are found and fixed.
___
### Pull request checklist

- [ ] Did you bump the project version?
_Not yet_
- [x] Did you add a description to the Pull Request?
- [x] Did you add new tests?
- [x] Did you add reviewers?

bountyHntr commented:
Thank you for contributing your PR! However, during testing I discovered some problems. 
Firstly, you use structures in your tests (`Foo`, `Foo2` etc.) whose names coincide with the names of structures from tests for deserialization from [`deser.jl`](https://github.com/bhftbootcamp/Serde.jl/blob/703f6a313c8bf51fb05eaf178d8aa73adabd5f66/test/deser.jl#L8). This results in errors like this:
```julia
Got exception outside of a @test
  invalid redefinition of constant Main.Foo2
  Stacktrace:
    [1] macro expansion
      @ ~/.julia/dev/Serde/test/deser.jl:40 [inlined]
    [2] macro expansion
      @ ~/.julia/juliaup/julia-1.10.0+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
    [3] macro expansion
      @ ~/.julia/dev/Serde/test/deser.jl:40 [inlined]
    [4] macro expansion
      @ ~/.julia/juliaup/julia-1.10.0+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
    [5] top-level scope
      @ ~/.julia/dev/Serde/test/deser.jl:7
    [6] include(fname::String)
      @ Base.MainInclude ./client.jl:489
    [7] top-level scope
      @ ~/.julia/dev/Serde/test/runtests.jl:10
    [8] include(mod::Module, _path::String)
      @ Base ./Base.jl:495
    [9] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:318
   [10] _start()
      @ Base ./client.jl:552
```
This error is also reported by the linter (VSCode in my case).
![Screenshot from 2024-02-14 19-57-03](https://github.com/bhftbootcamp/Serde.jl/assets/49916951/626e5abb-f75c-4acd-8bae-addbe89abdf8)
The test does not create a new scope, you can test this using the following simple example.
```julia
using Test

@testset "test_foo1" begin
    struct Foo
        a::Int
    end
end

@testset "test_foo2" begin
    struct Foo
        b::Int
    end
end

x = Foo(1)
```

bountyHntr commented:
I will also give an example from your test:
```julia
struct Foo2
    bool::Bool
    number::Float64
    nan::Float64
    inf::Float64
    _missing::Missing
    _nothing::Nothing
end

foo = Foo2(true, 163.163, NaN, Inf, missing, nothing)
println(Serde.to_yaml(foo))

# OUTPUT
bool: true
number: 163.163
nan: null
inf: null
_missing: null
_nothing: null
```
However, according to the [YAML specification](https://yaml.org/spec/1.2-old/spec.html#id2804092), for `NaN` and `Inf` it is necessary to use `.nan` and `.inf`, and not `null`, i.e. the result should be as follows:
```julia
bool: true
number: 163.163
nan: .nan
inf: .inf
_missing: null
_nothing: null
```

bountyHntr commented:
It's also a problem that you don't put quotes around strings when serializing them. Consider the following example:
```julia
julia> foo = Dict("v" => "0x0123")
Dict{String, String} with 1 entry:
  "v" => "0x0123"

julia> Serde.to_yaml(foo) |> println
v: 0x0123
```
This result corresponds to an integer and is equivalent to the one below:
```yaml
v: 291
```
At the same time the expected result:
```yaml
v: "0x0123"
```

This problem also occurs when using special characters, as in the example below:
```julia
julia> foo = Dict("v" => "\"Hello\"\'")
Dict{String, String} with 1 entry:
  "v" => "\"Hello\"'"

julia> Serde.to_yaml(foo) |> println
v: \"Hello\"'
```
This result is not equivalent to the expected one:
```yaml
v: "\"Hello\"'"
```
An equivalent result to the previous one can be obtained using `YAML.jl`:
```julia
julia> YAML.yaml(foo) |> println
: |-
  "Hello"'
```

maxfadson commented:
@bountyHntr Thanks for the extensive feedback!

Let's go through the points:
> The `@testset` does not create a new scope, you can test this using the following simple example.

I knew about this, but I completely forgot to run the entire test suite, my bad. Fixed.

> for NaN and Inf it is necessary to use .nan and .inf

That's right, I added the missing conditions to the function for numbers.

> In addition, when using types for which there are no corresponding ones from the YAML specification (for example, functions), you should just skip them or throw error

This problem covers the package as a whole, so far I have only added exception throwing for the "function" type, but this is clearly not the target solution. It might be worth discussing this in more detail and making it a new issue to develop a general solution.

> It's also a problem that you don't put quotes around strings when serializing them

I'm used to constantly seeing lines in Yaml without quotes and it turned out badly for me. I returned it and everything is fine now.

dmitrii-doronin commented:
```
name: "CI - Test Features"
on:
  push:
    branches:
      - main
  workflow_dispatch:

jobs:
  test:
    runs-on: ubuntu-latest
    continue-on-error: true
    strategy:
      matrix:
        features: [
            "anaconda",
            "aws-cli",
            "azure-cli",
            "common-utils",
            "conda",
            "desktop-lite",
            "docker-outside-of-docker",
            "docker-in-docker",
            "dotnet",
            "git",
            "git-lfs",
            "github-cli",
            "go",
            "hugo",
            "java",
            "kubectl-helm-minikube",
            "node",
            "nvidia-cuda",
            "oryx",
            "php",
            "powershell",
            "python",
            "ruby",
            "rust",
            "sshd",
            "terraform",
            "nix",
        ]
        baseImage:
          [
            "ubuntu:focal",
            "ubuntu:jammy",
            "debian:11",
            "debian:12",
            "mcr.microsoft.com/devcontainers/base:ubuntu",
            "mcr.microsoft.com/devcontainers/base:debian",
          ]
    steps:
      - uses: actions/checkout@v3

      - name: "Install latest devcontainer CLI"
        run: npm install -g @devcontainers/cli

      - name: "Generating tests for '${{ matrix.features }}' against '${{ matrix.baseImage }}'"
        run: devcontainer features test  --skip-scenarios -f ${{ matrix.features }} -i ${{ matrix.baseImage }} .

  test-scenarios:
    runs-on: ubuntu-latest
    continue-on-error: true
    strategy:
      matrix:
        features: [
            "anaconda",
            "aws-cli",
            "azure-cli",
            "common-utils",
            "conda",
            "desktop-lite",
            "docker-outside-of-docker",
            "docker-in-docker",
            "dotnet",
            "git",
            "git-lfs",
            "github-cli",
            "go",
            "hugo",
            "java",
            "kubectl-helm-minikube",
            "node",
            "nvidia-cuda",
            "oryx",
            "php",
            "powershell",
            "python",
            "ruby",
            "rust",
            "sshd",
            "terraform",
            "nix",
        ]
    steps:
      - uses: actions/checkout@v3

      - name: "Install latest devcontainer CLI"
        run: npm install -g @devcontainers/cli

      - name: "Testing '${{ matrix.features }}' scenarios"
        run: devcontainer features test -f ${{ matrix.features }} --skip-autogenerated  .

  test-global:
    runs-on: ubuntu-latest
    continue-on-error: true
    steps:
      - uses: actions/checkout@v3

      - name: "Install latest devcontainer CLI"
        run: npm install -g @devcontainers/cli

      - name: "Testing global scenarios"
        run: devcontainer features test --global-scenarios-only .
```

Can you try deserialising and then serialising the above example? It produces a valid yam but It seems that formatting is a bit off. Can you take a look into it?

gryumov commented on src/Ser/SerYaml.jl:
```diff
@@ -0,0 +1,293 @@
+const NEEDESCAPE =
+    Set{UInt8}(UInt8['"', '\\', '\b', '\f', '\n', '\r', '\t'])
+
+function escape_char(b)
```
We should consider refactoring this out 🤔

maxfadson replied:
makes sense, but I think it can be done as part of a separate issue

dmitrii-doronin replied:
Rewriting these pieces in a more straightforward manner would have been a boon

gryumov commented on src/Ser/SerYaml.jl:
```diff
@@ -0,0 +1,293 @@
+    yaml_value!(buf, x...; l = 1, is_root=true, kw...)
+    print(buf, "\n")
+    return String(take!(buf))
+end
```
```suggestion

end
```

maxfadson replied:
done

gryumov commented on src/Ser/SerYaml.jl:
```diff
@@ -0,0 +1,293 @@
+end
+
+function yaml_value!(buf::IOBuffer, f::Function, val::Number; kw...)::Nothing
+    if isnan(val) 
+        return print(buf, ".nan") 
+    end
+    if isinf(val) 
+        return print(buf, ".inf") 
+    end
+    return print(buf, val)
```
```suggestion
    return if isnan(val)
        print(buf, ".nan")
    elseif isinf(val)
        print(buf, ".inf")
    else
        print(buf, val)
    end
```

maxfadson replied:
I took this, it’s really more concise

maxfadson commented on src/Ser/SerYaml.jl:
```diff
@@ -0,0 +1,297 @@
+    global isroot = true
+    yaml_value!(buf, x...; l = 1, kw...)
+    print(buf, "\n")
+    return String(take!(buf))
```
fixed everything related to indents

maxfadson commented:
> Can you try deserialising and then serialising the above example? It produces a valid yam but It seems that formatting is a bit off. Can you take a look into it?

@dmitrii-doronin, I understand what you mean. I reconsidered the approach to handling newlines and applied it to the corresponding functions. now the document is serialized pretty :slightly_smiling_face:
I also added a test case for such cases, you can look it up.
